File: Y2021/Day14.java
package AdventOfCode.Y2021;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;

public class Day14 {

    static ArrayList<Character> polymer;
    static HashMap<String,Character> pairInsertions = new HashMap<>();
    static long[] countLetter;
    static HashMap<String,Long> countPairs;

    public static void parse() throws IOException {
        File f = new File("./AdventOfCode/Y2021/input14.txt");
        BufferedReader reader = new BufferedReader(new FileReader(f));
        String line;
        
        polymer = new ArrayList<>();
        countLetter = new long[26];
        countPairs = new HashMap<>();

        line = reader.readLine();
        for (char c : line.toCharArray()) polymer.add(c); 
        line = reader.readLine(); // skip empty line

        while ((line = reader.readLine()) != null) {
            String[] insertion = line.split(" -> ");
            pairInsertions.put(insertion[0], insertion[1].charAt(0));
        }

        reader.close();

        for (char c:polymer) countLetter[c-'A']++; // get initial atom counts
        
        for (int i=0; i<polymer.size()-1; i++) {
            String pair = "" + polymer.get(i) + polymer.get(i+1);
            countPairs.merge(pair, 1l, Long::sum); // log each pair count
        }
    }

    public static void getmax() {
        long min = Long.MAX_VALUE;
        long max = 0;
        for (long i: countLetter) {
            if (i>max) max=i;
            if (i<min && i>0) min=i;
        }
        
        System.out.println(max-min);
    }

    public static void part1() {
        for (int i=0; i<10; i++) {
            HashMap<String,Long> copy = (HashMap<String, Long>) countPairs.clone(); // take copy to not use new data while editing and to avoid concurrent modification
            copy.forEach((k,v) -> {
                long count = v;
                if (pairInsertions.containsKey(k)) {
                    char c = pairInsertions.get(k);
                    countLetter[c-'A'] += count;
                    countPairs.merge(k, -count, Long::sum); // [count] # of pairs are broken when an atom is added in the middle
                    countPairs.merge("" + k.charAt(0) + c, count, Long::sum); // [count] # of pairs are added with the atom to the left
                    countPairs.merge("" + c + k.charAt(1), count, Long::sum); // [count] # of pairs are added with the atom to the right
                }
            });
        }
        getmax();
    }

    public static void part2() {
        for (int i=0; i<40; i++) {
            HashMap<String,Long> copy = (HashMap<String, Long>) countPairs.clone(); // take copy to not use new data while editing and to avoid concurrent modification
            copy.forEach((k,v) -> {
                long count = v;
                if (pairInsertions.containsKey(k)) {
                    char c = pairInsertions.get(k);
                    countLetter[c-'A'] += count;
                    countPairs.merge(k, -count, Long::sum); // [count] # of pairs are broken when an atom is added in the middle
                    countPairs.merge("" + k.charAt(0) + c, count, Long::sum); // [count] # of pairs are added with the atom to the left
                    countPairs.merge("" + c + k.charAt(1), count, Long::sum); // [count] # of pairs are added with the atom to the right
                }
            });
        }
        getmax();
    }

    public static void main(String[] args) throws IOException {
        parse();        
        part1();

        parse();
        part2();
    }
}

File: 2022/day14.py
max_depth = 0
boundary = {}
def parse():
    global max_depth,boundary
    max_depth = 0
    boundary = {}
    with open("input14.txt") as f:
        for line in f:
            line = line.strip()
            line = line.split(" -> ")
            line = list(map(eval,line))
            depth = max(list(map(lambda x:x[1], line)))
            if depth > max_depth:
                max_depth = depth
            for i in range(1,len(line)):
                if line[i][0] == line[i-1][0]:
                    for j in range(line[i-1][1],line[i][1],1 if line[i-1][1] < line[i][1] else -1):
                        boundary[(line[i-1][0],j)] = True
                else:
                    for j in range(line[i-1][0],line[i][0],1 if line[i-1][0] < line[i][0] else -1):
                        boundary[(j,line[i-1][1])] = True

            boundary[line[-1]] = True

source = (500,0)

def part1():
    count = 0
    while True:
        sand = source
        while True:
            if sand[1] > max_depth:
                break
            if (sand[0],sand[1]+1) not in boundary:
                sand = (sand[0],sand[1]+1)
            elif (sand[0]-1,sand[1]+1) not in boundary:
                sand = (sand[0]-1,sand[1]+1)
            elif (sand[0]+1,sand[1]+1) not in boundary:
                sand = (sand[0]+1,sand[1]+1)
            else:
                boundary[sand] = True
                count += 1
                break
        if sand[1] > max_depth:
            break
    print(count)


def part2():
    count = 0
    while True:
        sand = source
        while True:
            if sand[1] == max_depth + 1:
                boundary[sand] = True
                count += 1
                break
            if (sand[0],sand[1]+1) not in boundary:
                sand = (sand[0],sand[1]+1)
            elif (sand[0]-1,sand[1]+1) not in boundary:
                sand = (sand[0]-1,sand[1]+1)
            elif (sand[0]+1,sand[1]+1) not in boundary:
                sand = (sand[0]+1,sand[1]+1)
            else:
                boundary[sand] = True
                count += 1
                break
        if source in boundary:
            break

    print(count)

parse()
part1()

parse()
part2()
